### ISSUE TITLE
Emotion example app updates entire tree every render (exceeding maximum update depth)

### ISSUE THREAD

**Issue by 3x071c** [remix-run/remix#2261]
### What version of Remix are you using?

1.2.3

### Steps to Reproduce

1. Download the [example app](https://github.com/remix-run/remix/tree/main/examples/emotion):
```console
$ svn export https://github.com/remix-run/remix/tree/main/examples/emotion
```
2. Install dependencies:
```console
$ pnpm upgrade # Minor version bumping before installation (doesn't matter for reproduction)
$ pnpm add -D @babel/core # Missing peer dependency (doesn't matter for reproduction)
```
3. Log style re-render trigger:
```console
$ vim app/entry.client.tsx # Add `console.log("🚨 RERENDER 🚨")` in `reset()` function body
```
4. Run app:
```console
$ pnpm dev
```
5. Look at the console:
![Browser console screenshot (clipped)](https://user-images.githubusercontent.com/87198856/157257628-be704bfd-a28b-45fb-a60a-56f120d03a11.png)

I believe this is not only re-applying the style tags, but also re-rendering the app over 50 times a second, which in either case isn't considered a good practice. Is there a way around this? I would recommend at least adding a warning to the emotion example for the time being. This can also be caught in CI with a proper ESLint configuration (`react/jsx-no-constructed-context-values`) :smile: 

### Expected Behavior

A performant app

### Actual Behavior

An app that is flooded with console warnings and re-renders

**Comment by PJColombo** [comment]
Hey @3x071c, can you share here how did you solve this issue :thinking: ?. 

Thanks!

**Comment by 3x071c** [comment]
@PJColombo If you pass an empty dependency array to `useEffect`, it will only re-apply the styles and re-render the entire app when `Document` is (un)mounted. To prevent unnecessary re-renders (around 3 on every navigation to an error boundary), I have memorized the children of the `Document` component (wrapping its JSX tree in a `useMemo` hook, and memorizing components with `React.memo`). This will still however re-apply the styles immediately after the page is first mounted, causing a performance hiccup on page reloads. I wrote a custom `useOnRemount` hook ([here](https://github.com/3x071c/lsg/blob/trunk/app/hooks/remount/useOnRemount.tsx)) to deal with that problem (you’ll have to use a context at the very top level to store a Boolean flag indicating whether the first mount has occurred or not). I’m still working on stopping `useEffect` from executing three times in a row when the Document is re-mounted by Remix. Re-opening as the example remains abad template for poorly performant apps and could be improved a lot.

**Comment by PJColombo** [comment]
I see. Thank you for the explanation @3x071c. All this performance problems keep users from using the template and makes the styled-component app example a better CSS-in-js remix alternative in my opinion. 

Let's keep the issue open until it get's solved or improved.

**Comment by tavoyne** [comment]
Why is all the boilerplate in the emotion example even necessary? It's just working out of the box for me.

**Comment by 3x071c** [comment]
@tavoyne [Docs](https://emotion.sh/docs/ssr)

**Comment by 3x071c** [comment]
@tavoyne Using emotion in the "out-of-the-box" SSR mode doesn't require an example, while the more advanced approach does. If one prefers the secondary approach for whatever reason (more control, better support etc.), I think they should get a proper example for that, just like Next.js (et al) do in their examples.

**Comment by tavoyne** [comment]
Thanks for clarifying that. Well, the only thing we have for now is a broken, contextless example. It could be wise to rename it `emotion-advanced`, and also to have an `emotion-basic` example. You mentioned Next.js, here's what you can read on emotion's website about it:

<img width="745" alt="image" src="https://user-images.githubusercontent.com/29731597/161934961-e044e5fd-7375-45be-b22f-02be980f1af2.png">

That kind of context is crucial for adoption, IMO.

Most people using emotion don't have a clue about how it works in the background.

**Comment by machour** [comment]
@tavoyne We'd be more than glad to review a PR with the changes you've mentionned!

**Comment by tavoyne** [comment]
@machour That's what I was planning to do. I just hoped that we could discuss things beforehand.

**Comment by 3x071c** [comment]
@tavoyne I don't think an `emotion-basic` example is necessary, as there's no configuration needed whatsoever. That's why most often the advanced SSR use case is being showcased in examples, because it's way harder to get right. I also feel like your suggestion is a little off-topic to mine; I just opened this to call attention to the problem with the current inefficient emotion example that some might mistakenly copy into their own projects as-is.

**Comment by tavoyne** [comment]
I'm not sure to agree. The point of examples is to help newcomers, not to confuse them. You mentioned Next.js, note that it does have a [_basic_ example](https://github.com/vercel/next.js/tree/canary/examples/with-emotion-vanilla) (which indeed is quite trivial) but has no _advanced_ one. About the fact that it's off topic, I apologise. I thought this was about making the emotion example better.

**Comment by 3x071c** [comment]
@tavoyne The basic example you linked actually does extract and inline critical CSS using "advanced" SSR techniques (definitely more simple than what the emotion docs suggest though). [This is the true basic example](https://github.com/vercel/next.js/tree/canary/examples/with-emotion), which as you can see is so stupidly simple (just install the dependencies and use the library) that there's no reason for one (but if you feel like it, you can of course submit a PR). This "basic" approach comes with **many** downsides that are not mentioned in the emotion docs and are the reason why so many users and maintainers reach for the "advanced" approach instead (see [these notes](https://github.com/mui/material-ui/issues/26561#issuecomment-855286153) by the official MUI maintainer, the [chakra-ui docs](https://chakra-ui.com/guides/getting-started/remix-guide) (a UI framework which uses emotion) etc.). I opened this in the hope of preventing more people from worsening their own app when getting chakra-ui/mui/emotion up and running with Remix, and I think providing a full example that prevents spec in compliance and other problems from arising should be the default/go-to approach in the Remix examples.

**Comment by AndlerRL** [comment]
I've been playing with the example for a brief time and saw this error too.

First, Emotion on SSR (as styled-components) injects all the styles after reading the requested components so the server understand what to render and finish the job. This means that the styles that we create it's being send to the server and then to the browser.

By knowing that, every time that we assign `emotionCache.sheet.container` we are executing funcs under the hood, because Emotion is trying to read your new styles and therefore it has to change and execute under the hood, making the main `emotionCache` to trigger once again and, since we need that for reading `emotionCache`, I thought that a guard might work.

The only instance that this is happening, it's on the client `useEffect` at `app/root.tsx`, so we take that principle in advantage and we verify if the HTMLNodeElement has new nodes/children or new siblings inside of children or node like:

```js
// Only executed on client
    useEffect(() => {
      // avoiding unwanted rendering
      if (document.head == emotionCache.sheet.container) return
// ...
```

At start it will take the first requested styles and, by testing this I haven't had problems with assigning new styles, removing styles/components removing nodes from function calls and going back/forward on the browser and styles are re-rendering whenever is necessary and performance issues were gone.

Hope not to miss something?

**Comment by ancashoria** [comment]
> 

I agree with @tavoyne. I just wanted to see if remix works with emotion and the first thing I did was to look for this example which I consider is the ideomatic way of integrating Emotion with Remix. It has confused me for sure.
